extensions/azurecore/src/account-provider/auths/azureAuth.ts

```diff
@@ -134,13 +157,130 @@ export abstract class AzureAuth implements vscode.Disposable {
 		}
 	}
 
-	public async hydrateAccount(token: Token, tokenClaims: TokenClaims): Promise<AzureAccount> {
-		const tenants = await this.getTenants(token.token);
-		const account = this.createAccount(tokenClaims, token.key, tenants);
+	private getHomeTenant(account: AzureAccount): Tenant {
+		// Home is defined by the API
+		// Lets pick the home tenant - and fall back to commonTenant if they don't exist
+		return account.properties.tenants.find(t => t.tenantCategory === 'Home') ?? account.properties.tenants[0] ?? this.commonTenant;
+	}
+
+	public async refreshAccess(account: AzureAccount): Promise<AzureAccount> {
+		// Deprecated account - delete it.
+		if (account.key.accountVersion !== AzureAuth.ACCOUNT_VERSION) {
+			account.delete = true;
+			return account;
+		}
+		try {
+			const tenant = this.getHomeTenant(account);
+			const tokenResult = await this.getAccountSecurityToken(account, tenant.id, azdata.AzureResource.MicrosoftResourceManagement);
+			if (!tokenResult) {
+				account.isStale = true;
+				return account;
+			}
+
+			return await this.hydrateAccount(tokenResult, this.getTokenClaims(tokenResult.token));
+		} catch (ex) {
+			if (ex instanceof AzureAuthError) {
+				void vscode.window.showErrorMessage(ex.message);
+				Logger.error(ex.originalMessageAndException);
+			} else {
+				Logger.error(ex);
+			}
+			Logger.error(ex);
+			account.isStale = true;
+			return account;
+		}
+	}
+
+	public async hydrateAccount(token: Token | AccessToken, tokenClaims: TokenClaims): Promise<AzureAccount> {
+		let account: azdata.Account;
+		if (this.authLibrary === 'ADAL') {
+			const tenants = await this.getTenants({ ...token });
+			account = this.createAccount(tokenClaims, token.key, tenants);
+		} else {
+			const tenants = await this.getTenantsMsal(token.token);
+			account = this.createAccount(tokenClaims, token.key, tenants);
+		}
 		return account;
 	}
 
-	protected abstract login(tenant: Tenant, resource: Resource): Promise<{ response: AuthenticationResult, authComplete: Deferred<void, Error> }>;
+	public async getAccountSecurityToken(account: AzureAccount, tenantId: string, azureResource: azdata.AzureResource): Promise<Token | undefined> {
+		if (account.isStale === true) {
+			Logger.error('Account was stale. No tokens being fetched.');
+			return undefined;
+		}
+
+		const resource = this.resources.find(s => s.azureResourceId === azureResource);
+		if (!resource) {
+			Logger.error('Invalid resource, not fetching', azureResource);
```

Comment:
Maybe something more like `Unable to find Azure resource ${azureResource} for account ${...} and tenant ${tenantId}`

In general make sure that all logging statements provide as much information as is reasonable and useful to have.